Propagate status field to cilium network policy

aanm opened the pull request:
Fixes #1222

`kubectl get ciliumnetworkpolicies multi-rules -o json`

```
...
    "status": {
        "Nodes": {
            "cilium-k8s-master": {
                "LastSeen": "2017-08-22T20:16:23.471822307-07:00",
                "Message": "\u003cnil\u003e",
                "OK": false
            },
            "cilium-k8s-node-2": {
                "LastSeen": "2017-08-22T20:15:59.540742656-07:00",
                "Message": "\u003cnil\u003e",
                "OK": false
            }
        }
    }

```

aanm committed:
k8s: moved CreateCustomResourceDefinitions to pkg/k8s

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Create CNP dedicated client with public interface

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
adding CRD cilium network policy policy status

When using kubernetes to manage CRD CNP, it was difficult to know on
which nodes was the policy being placed. With this commit it will be
possible to check with `kubectl get ciliumnetworkpolicies <name> -o
json` or with `kubectl describe ciliumnetworkpolicies <name>` when was
the policy received by which node.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -72,7 +72,7 @@ func (e *Endpoint) evaluateConsumerSource(owner Owner, ctx *policy.SearchContext
 	// Skip currently unused IDs
 	if ctx.From == nil || len(ctx.From) == 0 {
-		log.Debugf("[%s] Ignoring unused ID %v", e.PolicyID(), ctx)
+		//log.Debugf("[%s] Ignoring unused ID %v", e.PolicyID(), ctx)
```
Remove this for real and submit as separate PR

aanm replied:
Fixed

tgraf commented on pkg/k8s/third_party.go:
```diff
@@ -38,6 +40,45 @@ type CiliumNetworkPolicy struct {
+	Message string
+
+	// LastSeen contains the last time this status was updated
+	LastSeen time.Time
```
Let's call this `LastUpdated`

aanm replied:
Fixed

tgraf commented on pkg/k8s/third_party.go:
```diff
@@ -38,6 +40,45 @@ type CiliumNetworkPolicy struct {
+	OK bool
+
+	// Message describes the error condition if OK is false
+	Message string
```
Let's call this `Error`

aanm replied:
Fixed

aanm committed:
tests: adding e2e tests for CNP in kubernetes

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
 	}
+
+	rule.SetPolicyStatus(nodeaddress.GetName(), cnpns)
+	_, err = crdClient.Update(rule)
```
Move this status update into its own function in `pkg/k8s`

tgraf replied:
Will the status only work if CRD are in use?

aanm replied:
>Will the status only work if CRD are in use?

You got confused because of the variable name? It works for both, but I'll change the variable name to cnpClient

```
	// crdClient is the interface for CRD and TPR
	crdClient k8s.CRDInterface
```

tgraf replied:
Renaming will help. I'm also fine keeping it as-is as CRD will be the only interface soon and TPR will be gone. I was just wondering in general.

aanm replied:
Fixed

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
+		name := rule.Metadata.GetObjectMeta().GetName()
+		log.Warningf("k8s: unable to update CNP %s/%s with status: %s, retrying...", ns, name, err)
+		go func(ns, name string, rule *k8s.CiliumNetworkPolicy) {
+			for n := 0; n < 10; n++ {
```
Must be a const in `pkg/k8s/const.go` we should use a single value for all backoff loops so we can adjust globally.

aanm replied:
Fixed

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -932,11 +937,61 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
+					break
+				}
+				serverRule.SetPolicyStatus(nodeaddress.GetName(), cnpns)
+				_, err = crdClient.Update(serverRule)
```
Doesn't `Update()` give you the latest revision of the resource when it fails? You can use it instead of hoping for a race free get & update.

aanm replied:
>Doesn't Update() give you the latest revision of the resource when it fails? You can use it instead of hoping for a race free get & update.

It doesn't and it took me a while to figure out that. How can we still get a race with the get & update? I was hopping this would prevent it:
```
if serverRule.Metadata.UID != rule.Metadata.UID {
	// Although the policy was found this means it was deleted,
	// and re-added with the same name.
	log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
	break
}
```
but I can modify it to this as well:
```
if serverRule.Metadata.UID != rule.Metadata.UID || !serverRule.SpecEquals(rule)   {
	// Although the policy was found this means it was deleted,
	// and re-added with the same name.
	log.Debugf("k8s: rule %s/%s changed while updating node status, stopping retry", ns, name)
	break
}
```

tgraf replied:
The code is fine but it looked inefficient. It's unexpected that `Update()` does not provide the latest version.

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -963,6 +1018,22 @@ func (d *Daemon) deleteCiliumNetworkPolicy(obj interface{}) {
+		return
+	}
+
 	// FIXME
```
You can remove the FIXME here now

aanm replied:
Done

aanm committed:
daemon/k8s: wait for TPR resource definitions to be install in k8s

Since cilium was not checking if the TPR resource definition after being
created, it could cause false positives in the detection of conflicting
TPR and CRDs installed in the same cluster. With this commit, cilium
makes sure it TPR is installed before continue any operation.

Signed-off-by: André Martins <andre@cilium.io>
